src/webviews/extension-side/dataviewer/kernelDataViewerDependencyImplementation.ts

```diff
@@ -0,0 +1,107 @@
+
+'use strict';
+
+import { SemVer } from 'semver';
+import { ProductNames } from '../../../kernels/installer/productNames';
+import { Product } from '../../../kernels/installer/types';
+import { IApplicationShell } from '../../../platform/common/application/types';
+import { traceWarning } from '../../../platform/logging';
+import { DataScience, Common } from '../../../platform/common/utils/localize';
+import { EnvironmentType } from '../../../platform/pythonEnvironments/info';
+import { sendTelemetryEvent, Telemetry } from '../../../telemetry';
+import { executeSilently } from '../../../kernels/helpers';
+import { IKernel } from '../../../kernels/types';
+import { parseSemVer } from '../../../platform/common/utils';
+import { IDataViewerDependencyService } from './types';
+import { pandasMinimumVersionSupported } from './constants';
+
+export const kernelGetPandasVersion =
+    'import pandas as _VSCODE_pandas;print(_VSCODE_pandas.__version__);del _VSCODE_pandas';
+
+function kernelPackaging(kernel: IKernel): '%conda' | '%pip' {
+    const envType = kernel.kernelConnectionMetadata.interpreter?.envType;
+    const isConda = envType === EnvironmentType.Conda;
+    // From https://ipython.readthedocs.io/en/stable/interactive/magics.html#magic-pip (%conda is here as well).
+    return isConda ? '%conda' : '%pip';
+}
+
+/**
+ * Uses the Kernel to manage the dependencies of a Data Viewer.
+ */
+export class KernelDataViewerDependencyImplementation implements IDataViewerDependencyService {
+    constructor(private readonly applicationShell: IApplicationShell, private isCodeSpace: boolean) {}
+
+    protected async execute(command: string, kernel: IKernel): Promise<(string | undefined)[]> {
+        const outputs = await executeSilently(kernel.session!, command);
+        const error = outputs.find((item) => item.output_type === 'error');
+        if (error) {
+            traceWarning(DataScience.failedToGetVersionOfPandas(), error.message);
+        }
+        return outputs.map((item) => item.text?.toString());
+    }
+
+    protected async getVersion(kernel: IKernel): Promise<SemVer | undefined> {
+        try {
+            const outputs = await this.execute(kernelGetPandasVersion, kernel);
+            return outputs.map((text) => (text ? parseSemVer(text.toString()) : undefined)).find((item) => item);
+        } catch (e) {
+            traceWarning(DataScience.failedToGetVersionOfPandas(), e.message);
+            return;
+        }
+    }
+
+    private async installMissingDependencies(kernel: IKernel): Promise<void> {
+        sendTelemetryEvent(Telemetry.PythonModuleInstall, undefined, {
+            action: 'displayed',
+            moduleName: ProductNames.get(Product.pandas)!
+        });
+
+        let selection = this.isCodeSpace
+            ? Common.install()
+            : await this.applicationShell.showErrorMessage(
+                  DataScience.pandasRequiredForViewing().format(pandasMinimumVersionSupported),
```

Comment:
This is the code I'd like to avoid duplicating, comparing versions, displaying messages and the like.
Basically have a single class to handle dependency management and different implementations for how we get versions and install the missing packages.

Reply:
Note, tomorrow we could have another similar implementation for the debugger as well.

Reply:
> Note, tomorrow we could have another similar implementation for the debugger as well.

Yup, that's the idea!

Reply:
@sadasant just to check, your argument here is that we can get in one solid implementation here first, then take a second pass (once we add the second usage of it) to create a more general IInstaller? I don't think that I'm adverse to that, though it might mean that some of the choices we make now might be harder to generalize later.

Reply:
@IanMatthewHuff More or less. I'll say two arguments:

1. The debugger implementation might shake up our assumptions of what can be generalized and what can't.
2. The longer-term goal is not to have the installer logic in the dataViewerDependencyService, but in the `Installer`. Getting there means eventually merging the `IInstaller` with these `*DataViewerDependencyInstallers`.

Put in another way: The union between kernel+interpreter is likely going to be different than the union between kernel+interpreter+debugger, which is likely going to be different than the union between kernel+interpreter+debugger within `IInstaller`. Therefore, why try to formalize a more complex common API this early?

Reply:
(by "longer-term" I mean that my current focus is to get the DataFrame and the PlotViewer working, and then I'll resume with https://github.com/microsoft/vscode-jupyter/issues/10638 )

Reply:
> adverse to that, though it might mean that some of the choices we make now might be harder to generalize later.

I think this can be done now., 

> kernel+interpreter+debugger, which is likely going to be different than the union between

Not sure i agree about this, but anyways, for instance I can already see a lot of code duplication in this PR, but if this will be fixed in the next PR, then i'm fine with that.

Reply:
Once I have the picture clear with the debugger, the plan is to move all of the install code into the `IInstaller`. That will have to be fully generic for the three contexts: interpreter, kernel and debugger.